Training 2020/12

leogermond opened the pull request:
This training uses some new labs

+ typo corrections
+ general fixes to the generation script

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/AdaCore/training_material?pullRequest=9) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/AdaCore/training_material?pullRequest=9) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/AdaCore/training_material?pullRequest=9) it.</sub>

leogermond commented:
# Pandoc scripts

* Saner defaults (less verbose)
* Use the `--directories` arg as a default for pandoc's `--resource-path` (images...)
* Pandoc thin wrapper for generating static docs with a [default YAML file (like pandoc >v2.8)](https://pandoc.org/MANUAL.html#default-files)
* Pandoc new latex template for documents, based on the [eisvogel template](https://github.com/Wandmalfarbe/pandoc-latex-template) 

# Labs

* Radar basic lab: 3 Terminator-theme exercices using console wrapped behind a tiny API, for very simple declarations
* Solar System: historic AdaCore series of labs, based on the SDL, with an optional (and not tested by myself) STM32 binding

# CI

* Labs are built and packaged through GitHub actions

frank-at-adacore commented on courses/fundamentals_of_ada/labs/solar_system/06_subprograms/sol/main.adb.pp:
```diff
@@ -0,0 +1,204 @@
+-----------------------------------------------------------------------
```
This file looks like a prettyprint result - probably shouldn't be here

leogermond replied:
Good catch

frank-at-adacore commented on contrib/adacut:
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
```
shouldn't this file be "adacut.py"?

leogermond replied:
Ok for that. That's a hard to shake unix habit to hide exe extensions, but it will fail for windows users.

frank-at-adacore commented on extern/gnat_sdl/gnat_sdl/src/new_file:
```diff
@@ -0,0 +1,2 @@
+package body SDL_SDL_stdinc_h is
```
looks like a random extra file? (filename says "new_file")

leogermond replied:
Investigating, seems like a long-forgotten copy/paste issue indeed.

frank-at-adacore commented on pandoc/default_lab_docs.yaml:
```diff
@@ -0,0 +1,15 @@
+from: rst
```
This file is "yaml" but ".github/workflows/main.yml" is "yml" - consistency would be helpful (if we can)

leogermond replied:
Yes. It seems to be preferred convention to name them `.yaml` indeed

leogermond replied:
Oh I misunderstood. I will rename .github/workflows/main.yml -> yaml, it is supposed to work

frank-at-adacore commented on extern/game_support/stm32f4/README.txt:
```diff
@@ -0,0 +1,5 @@
+This example is used by the tutorial located in the GNAT Pro User�s Guide Supplement for Cross Platforms.  The tutorial is in the Annex titled "Tutorial: Embedded ARM Ada Project." 
```
Very small nit: Make this an MD file?

leogermond replied:
Yes. Also I can see some weird encoding issue (ISO-8859-1?) I'll have a look.

frank-at-adacore commented on support_files/templates/adacore.latex:
```diff
@@ -0,0 +1,1059 @@
+%%
+
+%%
+% This is the Eisvogel pandoc LaTeX template.
```
This is no longer the correct name.
Is there a reason we have both of these files here?

leogermond replied:
Simply copy/paste/fail. I kept the eisvogel as an eventual restart point, but it's not needed anymore, I'll simply document the source in the adacore template.

frank-at-adacore commented on courses/fundamentals_of_ada/labs/solar_system/18_multiple_inheritance/.~lock.questions.doc#:
```diff
@@ -0,0 +1 @@
+,leo,leo-Latitude-7410,03.12.2020 19:57,file:///home/leo/.config/libreoffice/4;
```
this file doesn't belong in the repository

leogermond replied:
Yes, removing and adding it to gitignore

frank-at-adacore reviewed (changes requested):
Solar_System labs need to be renumbered to match the sections in Fundamentals_of_Ada

I would prefer not having "obj" directories (and just use "." in the GPR files), but I see how they may be necessary with the "question_*" mechanism. Maybe we can refactor that before we merge?

Are NPP files for the STM32? Maybe we could put them in a subfolder, or something to make them "invisible" for the majority of the repository users?

To maintain consistency, I think every filename/folder should be all lowercase (if people see mixed case, they may create their own mixed-case, which could cause issues across multiple platforms)

Many RST files are DOS-formatted, which may cause issues / annoyance when being read on Linux

Solutions to "solar_system" labs should be in folder "answers" (which I've been using) or at least "solution" - you do know that "sol" is part of the solar system, right? :)

No way to add review comments to binary files, so:
   Files should be removed
      courses\fundamentals_of_ada\labs\solar_system\05_records\questions.doc 
      courses\fundamentals_of_ada\labs\solar_system\05_records\questions.pdf
      images\gnat_studio\2020-11-27_15-36.png
      images\gnat_studio\2020-11-27_15-38.png

   Why is images\title_background.pdf a PDF file instead of an image?


** NOTE **

I did not review the actual source files. I see that it's all added files, so I'm assuming they've been copied, and compile file, and work as expected

leogermond commented:
Long post, I've split it between what I agree with (at the end) and what needs to be discussed further :)

# To be discussed 

> Are NPP files for the STM32? Maybe we could put them in a subfolder, or something to make them "invisible" for the majority of the repository users?

I dont know I'll ask around. Moving them out of the way for the time being.

> Solutions to "solar_system" labs should be in folder "answers" (which I've been using) or at least "solution" - you do know that "sol" is part of the solar system, right? :)

Agreed. It's a bit of work to change it all, though. I'll try to see if a script can change them properly.
Is this blocking for merge ?

> I would prefer not having "obj" directories (and just use "." in the GPR files), but I see how they may be necessary with the "question_*" mechanism. Maybe we can refactor that before we merge?

It is not necessary for solutions, they both share the config and it could easily been changed later or on a case-by-case basis (using a scenario variable?).

I would disagree to changing that though. IMO the `obj/` directories is not simply a matter of taste.

1. a specific `obj/` directory can easily be `.gitignore` 'd, which is very useful considering we manipulate lot of files and GNAT generates many files as well. Ignoring these files gives us a clean `git status` so we're sure we handled all pending changes correctly. Or at least I am in my current workflow.
2. Also we want to hide most of this complexity ( `.ali`, `.bexch`) from the students, which an in-branch build wouldn't do.
3. Finally, it is standard to use the `obj/` directory in GNAT sources, and GNAT Studio automatically does that to new projects as well. 

So for these reasons I would be against this change if the argument for it is personal preference.

> Why is images\title_background.pdf a PDF file instead of an image?

It's the way the template handled it. The PDF is *very* lightweight independently of its render size because of the vector format. I could probably be converted to SVG, but I don't know much about this format, or Latex for that matter. So if this is an issue, I would keep it at least until some time after the merge.

# Agreed

> Solar_System labs need to be renumbered to match the sections in Fundamentals_of_Ada

Done

> To maintain consistency, I think every filename/folder should be all lowercase (if people see mixed case, they may create their own mixed-case, which could cause issues across multiple platforms)

I agree. I'll fix the image names.

> Many RST files are DOS-formatted, which may cause issues / annoyance when being read on Linux

Yes I believe I had a misconfigured git when I ported from svn. It was supposedly fixed, but I'll run a global dos2unix to make sure of that.

> No way to add review comments to binary files, so:
Files should be removed
courses\fundamentals_of_ada\labs\solar_system\05_records\questions.doc
courses\fundamentals_of_ada\labs\solar_system\05_records\questions.pdf
images\gnat_studio\2020-11-27_15-36.png
images\gnat_studio\2020-11-27_15-38.png

Thanks I'll be running and rebasing a removal commit.

> I did not review the actual source files. I see that it's all added files, so I'm assuming they've been copied, and compile file, and work as expected

Assuming is not necessary, the CI informs us that it compiles indeed. As for testing that it runs, it can be automatically tested as well. I manually tested them on both windows (with SDL) and linux (with SDL2) though.

leogermond commented:
It's mostly done now. I added `contrib/ci/check_images.py` to check the images being referenced and having a correct naming scheme (lowercase, extensions).

On that note, I added the rule that JPEG images should have the `.jpeg` extension only, and some of the images fail at that:
```
Invalid Type images/coffee_pot_runtime_error.jpg extension is invalid
Invalid Type images/spark_rightclick-source_tree.jpg extension is invalid
Invalid Type images/gnatprove-output-file.jpg extension is invalid
Invalid Type images/prove_dialog-basic.jpg extension is invalid
Invalid Type images/spark_rightclick-code.jpg extension is invalid
Invalid Type images/gnatstudio-preferences-spark.jpg extension is invalid
Invalid Type images/spark_menu-examine_file.jpg extension is invalid
Invalid Type images/gnatstudio-help-gnat.jpg extension is invalid
Invalid Type images/gnatprove-output-ide.jpg extension is invalid
Invalid Type images/prove_dialog-advanced.jpg extension is invalid
```

I propose that we don't fix those in this PR, as it would mean adding new files to the mix. Let's be flexible to keep things simple.

frank-at-adacore commented:
> > Solutions to "solar_system" labs should be in folder "answers" (which I've been using) or at least "solution" - you do know that "sol" is part of the solar system, right? :)
> 
> Agreed. It's a bit of work to change it all, though. I'll try to see if a script can change them properly.
> Is this blocking for merge ?

**_Not blocking for merge_**

> 
> > I would prefer not having "obj" directories (and just use "." in the GPR files), but I see how they may be necessary with the "question_*" mechanism. Maybe we can refactor that before we merge?
> 
> It is not necessary for solutions, they both share the config and it could easily been changed later or on a case-by-case basis (using a scenario variable?).
> 
> I would disagree to changing that though. IMO the `obj/` directories is not simply a matter of taste.

**_Understood. We can leave it (I just hate how GIT doesn't allow empty directories)_**

> 
> > Why is images\title_background.pdf a PDF file instead of an image?
> 
> It's the way the template handled it. The PDF is _very_ lightweight independently of its render size because of the vector format. I could probably be converted to SVG, but I don't know much about this format, or Latex for that matter. So if this is an issue, I would keep it at least until some time after the merge.

**_I always thought JPG was a better compression algorithm, but we can leave this for now_**

> 
> # Agreed
> > Solar_System labs need to be renumbered to match the sections in Fundamentals_of_Ada
> 
> Done
> 
> > To maintain consistency, I think every filename/folder should be all lowercase (if people see mixed case, they may create their own mixed-case, which could cause issues across multiple platforms)
> 
> I agree. I'll fix the image names.
> 
> > Many RST files are DOS-formatted, which may cause issues / annoyance when being read on Linux
> 
> Yes I believe I had a misconfigured git when I ported from svn. It was supposedly fixed, but I'll run a global dos2unix to make sure of that.
> 
> > No way to add review comments to binary files, so:
> > Files should be removed
> > courses\fundamentals_of_ada\labs\solar_system\05_records\questions.doc
> > courses\fundamentals_of_ada\labs\solar_system\05_records\questions.pdf
> > images\gnat_studio\2020-11-27_15-36.png
> > images\gnat_studio\2020-11-27_15-38.png
> 
> Thanks I'll be running and rebasing a removal commit.
> 
> > I did not review the actual source files. I see that it's all added files, so I'm assuming they've been copied, and compile file, and work as expected
> 
> Assuming is not necessary, the CI informs us that it compiles indeed. As for testing that it runs, it can be automatically tested as well. I manually tested them on both windows (with SDL) and linux (with SDL2) though.

frank-at-adacore commented:
I approved the changes, but didn't merge - you can do that either immediately or after whatever minor changes are still left

leogermond commented:
> always thought JPG was a better compression algorithm, but we can leave this for now

On "random" images it is. On large vector drawings, PDF or SVG are order of magnitude more efficient (they only describe the vertices information, not every pixel that compose them, unlike JPEG)

leogermond commented:
There was an issue on the generated files because of a bad renaming on my part, and I added the CI badge op top of the main README.md

leogermond commented:
The advanced labs renaming was not made for the inclusion in the courses:
```
Warning:  Could not load include file 'labs/advanced_privacy.lab.rst' at line 395 column 1
Warning:  Could not load include file 'labs/multiple_inheritance.lab.rst' at line 158 column 1
```
The labs PDF are not included in the zip anymore... I think I broke something on my side.

frank-at-adacore reviewed (approved):
Can you just write down a check-list in an email what needs to be fixed after merging this branch so we don't forget

leogermond commented:
Yes I will, I actually have most of that down as pending PR.
